vm/src/obj/objmemory.rs

```diff
@@ -1,74 +1,212 @@
-use super::objtype::{issubclass, PyClassRef};
-use crate::bytesinner::try_as_bytes;
+use std::{fmt::Debug, iter::FromIterator};
+
+use super::{objstr::PyString, objtype::PyClassRef};
+use crate::common::cell::{PyMappedRwLockReadGuard, PyMappedRwLockWriteGuard};
 use crate::common::hash::PyHash;
+use crate::obj::objbytes::{PyBytes, PyBytesRef};
+use crate::obj::objlist::{PyList, PyListRef};
 use crate::pyobject::{
     ItemProtocol, PyClassImpl, PyContext, PyObjectRef, PyRef, PyResult, PyValue, TypeProtocol,
 };
-use crate::slots::Hashable;
-use crate::stdlib::array::PyArray;
+use crate::slots::{AsBuffer, Hashable};
 use crate::VirtualMachine;
+use crate::{bytesinner::try_as_bytes, pyobject::IntoPyObject};
+use crossbeam_utils::atomic::AtomicCell;
+
+pub trait BufferProtocol: Debug + Sync + Send {
```

Comment:
Instead of `Send + Sync` here, you probably want to use `PyThreadingConstraint` from `pyobject`